### ISSUE TITLE
Add Entity class to CSharpFunctionExtensions

### ISSUE THREAD

**Issue by robertlarkins** [vkhorikov/CSharpFunctionalExtensions#207]
This library has the ValueObject class, but not an Entity class, like what is available here:
https://github.com/vkhorikov/DddAndEFCore/blob/master/src/App/Entity.cs
is the Entity class likely to be added?

**Comment by vkhorikov** [comment]
I haven't added it to the library because of this line here: https://github.com/vkhorikov/DddAndEFCore/blob/master/src/App/Entity.cs#L75

It's a concern specific to EF Core, which makes this class non-universal (unlike ValueObject). If you use NHibernate, then this method will look differently: https://github.com/vkhorikov/CqrsInPractice/blob/master/After/src/Logic/Students/Entity.cs#L48-L51

**Comment by robertlarkins** [comment]
Ah I see, and I suppose that line would differ per ORM. Shall we leave this issue/question open, or close it?

**Comment by vkhorikov** [comment]
Yes, the difference per ORM and also when you don't use any ORM. I'll close the issue.

**Comment by robertlarkins** [comment]
@vkhorikov As the ValueObject class now has code specific to EF Core (#224), is there an argument to made that an Entity class could be added to the library? I know that it means it has a reference to an ORM (or ORMs), but could it check for an object's real or unproxied type for EF Core then NHibernate, and if it isn't proxied then use the object type?

Maybe something like this (untested)
```
private Type GetRealType()
{
	Type type = GetType();

	if (type.ToString().Contains("Castle.Proxies."))
		return type.BaseType;

	return NHibernateProxyHelper.GetClassWithoutInitializingProxy(this);
}
```

**Comment by vkhorikov** [comment]
Yes, I was contemplating on this too after I added that code. The unproxying code for NHibernate can be done the same way as for EF Core, without the need to refer to `NHibernateProxyHelper`. I'm reopening this issue, will experiment with it a next week.

**Comment by vkhorikov** [comment]
Entity implementation is available in 2.12.0
